docs/0.19-changes.md

```diff
@@ -0,0 +1,93 @@
+
+## Rationale behind changes
+
+The beacon API has changed significantly since it was first released.  These changes generally fall into three categories:
+
+- additional query parameters
+- additional response metadata in JSON and/or headers
+- introduction of new encoding systems, notably SSZ, for some endpoints
+
+Due to these changes, it has become increasingly difficult to provide full functionality without changing function signatures.  And, if function signatures are going to change, it makes sense to change them once in a more sustainable fashion rather than have a continual drip of function signature changes each time the beacon API changes.  0.19.0 brings these signature changes, and 
+
+## Details of changes and migration guidance
+
+### Consolidation of call options
+Many endpoints in the beacon API have acquired additional parameters since the first release of the API specification.  To allow for current options and to ease addition of future options all options have been moved to function-specific structs.  For example, a call to fetch validators used to look something like this:
+
+```go
+validatorsProvider.Validators(ctx, "head", nil)
+```
+
+Even with a relatively simply call like this it is not clear what these parameters represent, especially when one is `nil`.  With the consolidation of call options this now looks something like this:
+
+```
+validatorsProvider.Validators(ctx, &api.ValidatorsOpts{
+  State: "head",
+})
+```
+
+Which makes it clearer that the `"head"` parameter refers to the state, and hides the explicit `nil` parameter for the specific validator indices.  Additional parameters added to the API can then be added to the struct without further disrupting the function signature.
+
+### Relocation of `Error` struct and extended usage
+`http.Error` has moved to `api.Error`, to make it clearer that it is reporting an error found within the API rather than specifically an HTTP error.
+
+Not all errors returned by the call will be `api.Error`, only those that return an error relating to the server.  So, for example, an attempt to call a function with a `nil` options struct (see below for options structures) will return a simple error.  But when an `api.Error` is available it will provide additional information about why the API rejected a request.
+
+### Response struct
+Responses from endpoints are now wrapped in the `api.Response` struct.  The requested information can be found in the `Data` element of this struct, and any metadata supplied in the response can be found in the `Metadata` element of this struct.  Metadata has been added to a number of endpoints to support situations where additional data may be useful to the user but is not part of the requested structure.  For example:
+
+```go
+response, _ := beaconBlockProvider.SignedBeaconBlock(ctx, &api.SignedBeaconBlockOpts{
+  Block: "123",
+})
+
+finalized, exists := response.Metadata["finalized"]
```

Comment:
I do think it would be nice to provide constants for known metadata keys, seeing as they are defined in the spec. It would help avoid uncertainty around `executionOptimistic` vs `execution_optimistic` etc

Reply:
Added constants in `api/metadata`.